flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/FlinkAggregateExpandDistinctAggregatesRule.java

```diff
@@ -471,10 +474,24 @@ private void rewriteUsingGroupingSets(RelOptRuleCall call,
 			final RexNode nodeZ = nodes.remove(nodes.size() - 1);
 			for (Map.Entry<ImmutableBitSet, Integer> entry : filters.entrySet()) {
 				final long v = groupValue(fullGroupSet, entry.getKey());
-				nodes.add(
+				// Remap and get the filterArg of the distinct aggregate call.
+				int distinctAggCallFilterArg = remap(fullGroupSet,
+					groupSetToDistinctAggCallFilterArg.getOrDefault(entry.getKey(), -1));
+				if (distinctAggCallFilterArg < 0) {
+					nodes.add(
 						relBuilder.alias(
-								relBuilder.equals(nodeZ, relBuilder.literal(v)),
-								"$g_" + v));
+							relBuilder.equals(nodeZ, relBuilder.literal(v)),
+							"$g_" + v));
+				} else {
+					RexBuilder rexBuilder = aggregate.getCluster().getRexBuilder();
+					// merge the filter of the distinct aggregate call itself.
+					nodes.add(relBuilder.alias(
```

Comment:
it's better to extract `nodes.add(relBuilder.alias(project, "$g_" + v)` as common part